documentation/design/proposals/kafka-address-space.adoc

```diff
@@ -0,0 +1,177 @@
+
+* kafka.spec.listeners
+
+When a tenant creates a kafka address space, the `address-space-controller` will use the `KafkaConfig` resource to create a new `Kafka` Strimzi resource using a name based on the address space name and namespace.
+
+==== Endpoints
+
+For the Kafka address space, the `AddressSpace` endpoints provides the endpoints of the Kafka bootstrap servers. An example address space for a `kafka` address space type that has been created may look like this:
+
+[source,yaml,options="nowrap",subs="attributes"]
+----
+apiVersion: enmasse.io/v1beta1
+kind: AddressSpace
+metadata:
+  name: myspace
+spec:
+  type: kafka
+  plan: kafka-small
+  endpoints:
+  - name: kafka
+    service: kafka
+    cert:
+      provider: certBundle
+      tlsKey: // Base64-encoded string
+      tlsCert: // Base64-encoded string
+   expose:
+     type: route
+status:
+  endpointStatuses:
+  - name: kafka
+    serviceHost: kafka-bootstrap-6bf3d9b.enmasse-infra.svc
+----
+
+The implementation will attempt to map the endpoint specification to the `Kafka` resource listener definitions. 
+
+Kafka clients may only authenticate using SASL in the initial version of the Kafka address space. 
+
+==== Authentication services
+
+Kafka integrates with authentication services by adding support for SASL forwarding/plugin to Kafka. EnMasse will automatically configure the Kafka cluster with information how to connect to the authentication service.
+
+
+==== Addresses
+
+EnMasse `Address` resources maps to `KafkaTopic` resources using the `topic` address type, and `AddressPlan` resources describing the amount of the cluster a topic uses. Address properties can be specified as part of the `AddressPlan` or the `Address`. The simplest form is to use defaults and settings from the plan:
+
+[source,yaml,options="nowrap",subs="attributes"]
+----
+apiVersion: enmasse.io/v1beta1
+kind: Address
+metadata:
+  name: myspace.mytopic
+spec:
+  type: topic
+  plan: kafka-topic-small
+----
+
+The `AddressPlan` may look like this:
+
+[source,yaml,options="nowrap",subs="attributes"]
+----
+apiVersion: admin.enmasse.io/v1beta2
+kind: AddressPlan
+metadata:
+  name: kafka-topic-small
+spec:
+  addressType: topic
+  resources:
+    kafka: 0.01
+  allowPropertiesOverride: true
+  properties:
+    partitions: 1
+    replicas: 2
+     retention.ms: 10000
+     segment.bytes: 65536
+----
+
+The `api-server` will create a Strimzi `KafkaTopic` based on the `Address` and `AddressPlan`.
+
+A more advanced `Address` where a greater set of Kafka topic features are exposed:
+
+[source,yaml,options="nowrap",subs="attributes"]
+----
+apiVersion: enmasse.io/v1beta1
+kind: Address
+metadata:
+  name: myspace.mytopic
+spec:
+  type: topic
+  plan: kafka-topic-small
+  properties:
```

Comment:
What I don't like with the "bag of properties" concept here, is that you no real way to validate that. I would suggest to follow the pattern of Kubernetes instead: create dedicated types for the different choices: see https://github.com/kubernetes/api/blob/master/core/v1/types.go#L49

Reply:
@ctron, I’m not sure I understand the reasoning behind distinct types leading to better validation.   

The proposal is not describing a “bag of properties” without a schema. Each field within the properties object may have different and possibly nested types.

This should provide the same level of validation of each property as you would from having separate objects for different address types.

Moreover, any such object would need to allow any property to be omitted and have a default in order to simplify configuration. In addition, the implementation would need to allow unknown properties in order to handle forward and backward compatibility.

With the above requirements I think either direction will give the same benefits. 

A challenge in having separate fields would be that we have the same types (I.e topic) defines in multiple address space types, so i don’t think distinct properties fields for each type makes sense.

If this makes sense to you, I can update the proposal to clarify this bit.

Reply:
So I think the conclusion is to use the `properties` field name, but provide OpenAPI spec to ensure that only a valid set of properties are specified together. I will update this proposal to reflect that, does that sound ok @ctron ?